### PR TITLE
Overflow set render props

### LINKED ISSUE #4081: OverflowSet.types.d.ts shouldn't use any for its types
### Bug Report

- __Package version(s)__: 5.5.2

#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? Yes

Requested priority: Normal 

Products/sites affected: Sway

#### Describe the issue:

In `OverflowSet.types.d.ts`:

```typescript
/**
 * Rendering method for overflow button and contextual menu. The argument to the function is
 * the overflowItems passed in as props to this function.
 */
onRenderOverflowButton: IRenderFunction<any[]>;
```

This doesn't play well with TypeScript's `--strictFunctionTypes`, as that flag won't allow passing functions with more specific types than `any[] | undefined` as the render function. The type it takes in should instead be `IOverflowSetItemProps[]`.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #4081
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Used specific `OverflowSet` types for `onRenderOverflowButton` props instead of `any`.

### CODE REVIEW

**Comment by micahgodbolt** [comment]
How about IButtonProps instead? It's supposed to be a button with contextual menu, which IButtonProps 100% supports. Or do you want to render something other than a button there?

File: packages/office-ui-fabric-react/src/components/OverflowSet/OverflowSet.types.ts [hunk 1]
```diff
@@ -46,7 +46,7 @@ export interface IOverflowSetProps extends React.Props<OverflowSet> {
    * Rendering method for overflow button and contextual menu. The argument to the function is
    * the overflowItems passed in as props to this function.
   */
-  onRenderOverflowButton: IRenderFunction<IOverflowSetItemProps[]>;
+  onRenderOverflowButton: IRenderFunction<IButtonProps[]>;
```
Comment (micahgodbolt) [inline packages/office-ui-fabric-react/src/components/OverflowSet/OverflowSet.types.ts]: oops, didn't notice the array notation. overflowButton is singular, so it'd just be 1 button.

Comment (JoshuaKGoldberg) [reply packages/office-ui-fabric-react/src/components/OverflowSet/OverflowSet.types.ts]: I'm confused. Why was it `any[]` and not `any` in the first place?

**Comment by micahgodbolt** [comment]
Ah shoot. I take it back. it's the overflow items that are passed into the render function.

that'd be IContextualMenuItem[]

You're right. It was an array before. Now realizing that the only props passed to the function were the items.

**Comment by micahgodbolt** [comment]
good reason to have proper typings :D

**Comment by micahgodbolt** [comment]
How about updating the type in the OverflowSet.tsx as well

```
 private _onRenderOverflowButtonWrapper(items: any[]): JSX.Element {

**Comment by micahgodbolt** [comment]
might be a case where you say "as IContextualMenuItems[]" but i'm not sure. Give it a shot

**Comment by dzearing** [comment]
I'm concerned with this. Can we do something like `Object` or maybe `{}`? Ideally this would be a generic but they don't mix well with tsx.

Basically I want any prop type to be passed in and work, and this seems somewhat limiting I would think.

**Comment by JoshuaKGoldberg** [comment]
`{}` works for me.

**Comment by micahgodbolt** [comment]
@JoshuaKGoldberg there are SEVERAL areas that we'd like to implement better generic support. Lots of instances where we pass in custom render functions along with arrays of data to be rendered.

**Comment by dzearing** [comment]
@JoshuaKGoldberg I pushed a little too soon. I reverted because I believe this could be a breaking change.

**Pull request merged**